adr/0008-named-callbacks.md

```diff
@@ -0,0 +1,55 @@
+# 7: LogAs Function in Capability Class "Named Callbacks"
+
+Date: 2024-03-21
+
+## Status
+
+Proposed
+
+## Context
+
+Pepr utilizes anonymous callbacks that produce logs during execution. This approach, while functional, has led to challenges in debugging and testing. The logs generated by these callbacks lack clear identifiers, making it difficult to ascertain which callback produced which log unless manual names are embedded in the log messages from within the callbacks themselves.
+
+
+## Decision Drivers
+
+The key drivers for this decision include:
+
+- **Devex**: Instead of manual declaration of `const name = "my-action";` in each callback and each log, we can add a name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`
+- **Debugging Efficiency**: The debugging process in the current system is time-consuming and error-prone due to the anonymous nature of callbacks.  
+- **Code Readability**: Anonymous callbacks can make the code harder to understand, especially for new developers or when returning to old code that has a different form of naming.
+- **Maintenance and Scalability**: As the system grows, the manageability of anonymous callbacks becomes increasingly burdensome to maintain a consistent state of naming across actions.
+- **Error Tracing**: Pinpointing the source of errors or unexpected behavior in the system is complex without clear identification of each callback.
+
+
+## Considered Options
+
+1. **Named Callbacks**: Add an optional name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`.
+2. **Maintain Anonymous Callbacks**: Continue using anonymous callbacks without any changes and move this responsibility to the module author. 
+
+## Decision Outcome 
+
+Chosen option: "Named Callbacks", because it provides a clear and consistent way to identify the source of logs and makes the code easier to understand, debug, and test with an easy developer experience.
+
+```ts
+// Watch for changes to the UDSPackage CRD to enqueue a package for processing
+When(UDSPackage)
+  .IsCreatedOrUpdated()
+  // Advanced CR Validation
+  .Validate(validator)
+  // Enqueue the package for processing
+  .Reconcile(reconciler)
+  .LogAs("UDS Operator Reconciler")
+```
+
+
+## Benefits
+
+- Simpler developer experience by reducing workload on the module author.
+- Improved Logs and Debugging by programmatically identifying the source of logs in a repeatable fashion across modules.
+
+## Consequences
+
+1. Add another optional function to the Capability class, `LogAs`, to add a name to the callback.
```

Comment:
```suggestion
1. Add another optional function to the Capability class, `LogAs`, to add a name to the callback (and also maybe add a default, generated name for Pepr to use when a dev forgets to set one).
```

Or is this too implementation focused?

Reply:
Agreed we should figure out a default name.